### DISCUSSION TITLE
Insert statement silently drops data when passed variable sets of columns

### DISCUSSION THREAD

**Discussion by ucwillg** [sqlalchemy/sqlalchemy#11140] [Usage Questions]
I'm opening a discussion because this could plausibly be intended behavior. However, it was surprising to me and should potentially be documented if it is.

When using an insert statement on a table, if i pass VALUES in that have differing sets of columns, i found that the emitted insert statement would silently drop values. It seems as if a set of columns is compiled from the keys of the first object in the array, and the others are required to have the same set.

Here's a script to replicate this:
```
import sys

import sqlalchemy
from sqlalchemy import insert
from sqlalchemy import Column, Integer, create_engine
from sqlalchemy.orm import declarative_base, sessionmaker


def main(postgres_url: str) -> None:
    print("Version:", sqlalchemy.__version__)
    # Update the connection string with your PostgreSQL credentials and database name
    engine = create_engine(postgres_url, echo=True)

    Base = declarative_base()

    class TestTable(Base):
        __tablename__ = "test_table"
        id = Column(Integer, primary_key=True)
        a = Column(Integer)
        b = Column(Integer)

    Base.metadata.create_all(engine)

    Session = sessionmaker(bind=engine)
    with Session() as session:
        session.query(TestTable).delete(synchronize_session=False)
        session.commit()

    A = 1
    B = 2
    C = 3

    with Session() as session:
        row_a = {
            TestTable.a.key: A,
        }
        row_b = {
            TestTable.a.key: B,
            TestTable.b.key: C,
        }
        statement = insert(TestTable).values([row_a, row_b])
        session.execute(statement)
        session.commit()

    with Session() as session:
        for row in session.query(TestTable):
            print(row.__dict__)

if __name__ == "__main__":
    main(sys.argv[1])
```

Here's the output:

```
Version: 2.0.25
2024-03-11 18:52:02,348 INFO sqlalchemy.engine.Engine select pg_catalog.version()
2024-03-11 18:52:02,348 INFO sqlalchemy.engine.Engine [raw sql] {}
2024-03-11 18:52:02,349 INFO sqlalchemy.engine.Engine select current_schema()
2024-03-11 18:52:02,349 INFO sqlalchemy.engine.Engine [raw sql] {}
2024-03-11 18:52:02,350 INFO sqlalchemy.engine.Engine show standard_conforming_strings
2024-03-11 18:52:02,350 INFO sqlalchemy.engine.Engine [raw sql] {}
2024-03-11 18:52:02,350 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-03-11 18:52:02,352 INFO sqlalchemy.engine.Engine SELECT pg_catalog.pg_class.relname
FROM pg_catalog.pg_class JOIN pg_catalog.pg_namespace ON pg_catalog.pg_namespace.oid = pg_catalog.pg_class.relnamespace
WHERE pg_catalog.pg_class.relname = %(table_name)s AND pg_catalog.pg_class.relkind = ANY (ARRAY[%(param_1)s, %(param_2)s, %(param_3)s, %(param_4)s, %(param_5)s]) AND pg_catalog.pg_table_is_visible(pg_catalog.pg_class.oid) AND pg_catalog.pg_namespace.nspname != %(nspname_1)s
2024-03-11 18:52:02,352 INFO sqlalchemy.engine.Engine [generated in 0.00013s] {'table_name': 'test_table', 'param_1': 'r', 'param_2': 'p', 'param_3': 'f', 'param_4': 'v', 'param_5': 'm', 'nspname_1': 'pg_catalog'}
2024-03-11 18:52:02,355 INFO sqlalchemy.engine.Engine COMMIT
2024-03-11 18:52:02,356 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-03-11 18:52:02,356 INFO sqlalchemy.engine.Engine DELETE FROM test_table
2024-03-11 18:52:02,356 INFO sqlalchemy.engine.Engine [generated in 0.00007s] {}
2024-03-11 18:52:02,357 INFO sqlalchemy.engine.Engine COMMIT
2024-03-11 18:52:02,358 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-03-11 18:52:02,359 INFO sqlalchemy.engine.Engine INSERT INTO test_table (a) VALUES (%(a_m0)s), (%(a_m1)s)
2024-03-11 18:52:02,359 INFO sqlalchemy.engine.Engine [no key 0.00008s] {'a_m0': 1, 'a_m1': 2}
2024-03-11 18:52:02,360 INFO sqlalchemy.engine.Engine COMMIT
2024-03-11 18:52:02,361 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-03-11 18:52:02,361 INFO sqlalchemy.engine.Engine SELECT test_table.id AS test_table_id, test_table.a AS test_table_a, test_table.b AS test_table_b
FROM test_table
2024-03-11 18:52:02,362 INFO sqlalchemy.engine.Engine [generated in 0.00007s] {}
{'_sa_instance_state': <sqlalchemy.orm.state.InstanceState object at 0x10f1707d0>, 'b': None, 'id': 11, 'a': 1}
{'_sa_instance_state': <sqlalchemy.orm.state.InstanceState object at 0x10f170830>, 'b': None, 'id': 12, 'a': 2}
2024-03-11 18:52:02,362 INFO sqlalchemy.engine.Engine ROLLBACK
```

You can see that the value for "b" was not emitted in sql.

It seems like the value made it as far as the base insert class instance:

```
> /Users/willgoldie/main/replicate.py(43)main()
     42         import ipdb; ipdb.set_trace()
---> 43         session.execute(statement)
     44         session.commit()

ipdb> statement
<sqlalchemy.sql.dml.Insert object at 0x10ce5c3d0>
ipdb> statement._multi_values
([{Column('a', Integer(), table=<test_table>): 1}, {Column('a', Integer(), table=<test_table>): 2, Column('b', Integer(), table=<test_table>): 3}],)
ipdb>
```

Caveats:
- I only tested in postgresql.
- bulk_insert_mappings works properly for this case, in my testing.

**Comment by zzzeek** [comment]
hi -

What SQL would you expect to see for this statement?


```py
        statement = insert(TestTable).values([
                {TestTable.a.key: 1}
                {TestTable.a.key: 1, TestTable.b.key: 2}
        ])
```


I would most prefer that raised an error, however as this is for bulk purposes we are trying to conserve on python overhead for every row.    there might be more explicit errors when using execute(stmt,params) but I would have to check.

**Reply by ucwillg** [reply]
Hi, thank you for the response.

I don't have a strong opinion on the best behavior. An error or warning would be fine for our use cases, less surprising, and would have avoided the bug that motivated this question. 

I saw that bulk update mappings emits multiple queries, which is less surprising to me, but admittedly not the behavior we'd aim for our application to have in practice (we'll probably fill nulls into the mappings most cases; and use heterogenous parameter dictionaries in other cases). I suppose something like render_nulls (render_missing?) that emits a single insert with the union of all columns in the insert could be a convenience feature, but ultimately something we can build on top of the existing interfaces.

**Reply by zzzeek** [reply]
we get the number of values from the first entry.    we then try to spend as little time as possible on the remaining dictionaries.   if there are 1000 dictionaries, that would be 1000 more len() calls, so I'm not too enthused about adding that in.    is your code unit tested?

**Reply by ucwillg** [reply]
Yeah that's fine, it's obviously your call where the balance between safety and performance should live for the interface, and we can build the behavior we need in different cases regardless. 

Would you be open to a PR documenting this nuance, e.g at https://docs.sqlalchemy.org/en/20/core/dml.html#sqlalchemy.sql.expression.Insert.values?

>  is your code unit tested?

The use of unit tests seems orthogonal to this discussion. In this case, we in did in fact discover this behavior through a test failing. It still consumed time and attention.

**Reply by zzzeek** [reply]
it definitely should be documented, I had looked and I had assumed there must be doc warnings by now, maybe they are elsewhere for the `.execute(stmt, params)` case.    I'm not sure if we might have some handling for the `.execute(stmt, params)` case also.    The same approach should likely be used for both (similar doc warnings , if there are assertions in the code they should be in both places).

I almost wonder if there's a way to get pep-484 typing to figure this out also.

**Reply by CaselIT** [reply]
> I almost wonder if there's a way to get pep-484 typing to figure this out also.

no, that would need the ability of pep484 do define typed dict dynamically. Also typed dict can only have strings as keys, so that's a non starter in any case here
